File: src/main/java/org/cubexmc/railway/physics/LeashedRailPhysics.java
package org.cubexmc.railway.physics;

import org.cubexmc.railway.model.Stop;
import org.cubexmc.railway.train.TrainInstance;

/**
 * Leashed mode: reactive movement + visual leash couplers.
 */
public class LeashedRailPhysics extends ReactiveRailPhysics {

    private LeashCoupler coupler;

    @Override
    public void init(TrainInstance train) {
        super.init(train);
        coupler = new LeashCoupler(train.getService().getPlugin(), train);
        coupler.start();
    }

    @Override
    public void onDeparture(TrainInstance train, Stop fromStop) {
        super.onDeparture(train, fromStop);
        if (coupler == null) {
            coupler = new LeashCoupler(train.getService().getPlugin(), train);
            coupler.start();
        }
    }

    @Override
    public void tick(TrainInstance train, double timeFraction, long currentTick) {
        // Physics settings (setSlowWhenEmpty, setGravity, etc.) are handled by
        // super.tick()
        super.tick(train, timeFraction, currentTick);
        if (coupler != null)
            coupler.update();
    }

    @Override
    public void cleanup(TrainInstance train) {
        if (coupler != null) {
            coupler.cleanup();
            coupler = null;
        }
        super.cleanup(train);
    }
}

File: src/main/java/org/cubexmc/metro/physics/LeashedRailPhysics.kt
package org.cubexmc.metro.physics

import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.train.TrainInstance

/** Leashed mode: reactive movement + visual leash couplers. */
open class LeashedRailPhysics : ReactiveRailPhysics() {

    private var coupler: LeashCoupler? = null

    override fun init(train: TrainInstance) {
        super.init(train)
        val createdCoupler = LeashCoupler(train.service.plugin, train)
        coupler = createdCoupler
        createdCoupler.start()
    }

    override fun onDeparture(train: TrainInstance, fromStop: Stop?) {
        super.onDeparture(train, fromStop)
        if (coupler == null) {
            val createdCoupler = LeashCoupler(train.service.plugin, train)
            coupler = createdCoupler
            createdCoupler.start()
        }
    }

    override fun tick(train: TrainInstance, timeFraction: Double, currentTick: Long) {
        // Physics settings (setSlowWhenEmpty, setGravity, etc.) are handled by super.tick().
        super.tick(train, timeFraction, currentTick)
        coupler?.update()
    }

    override fun cleanup(train: TrainInstance) {
        coupler?.cleanup()
        coupler = null
        super.cleanup(train)
    }
}
